### PR TITLE
fixes and new sensors

### PR DESCRIPTION
**Pull request by jrgutier**
Fixes:
Save last known good JSON so if the server blips sensors won't go unavailable

Added Sensors:
3rd Row Left Seat Heat Status
3rd Row Right Seat Heat Status
Defrost/Defog Status
EMS Charger EVSE Type
Estimated Cabin Temperature
Left Seat Heat Status
Left Seat Vent Status
Offline Mode Enabled
Rear Left Seat Heat Status
Rear Right Seat Heat Status
Remote Charging Available
Right Seat Heat Status
Right Seat Vent Status
Set Temperature Mobile Status
SOC Limit
Steering Wheel Heat Status
Vehicle Charger Status
Window Calibration Front Left State
Window Calibration Front Right State
Window Calibration Rear Left State
Window Calibration Rear Right State

### CODE REVIEW

File: custom_components/rivian/const.py [hunk 1]
```diff
@@ -306,16 +298,140 @@
...
+    ),
+    "thermal/hvac_mobile_status/right_seat_heat_status": RivianSensorEntity(
+        entity_description=RivianSensorEntityDescription(
+            name="Right Seat Heat Status",
+            key=f"{DOMAIN}_thermal_hvac_mobile_status_right_seat_heat_status",
+        )
+    ),
+    "thermal/hvac_mobile_status/right_seat_vent_status": RivianSensorEntity(
+        entity_description=RivianSensorEntityDescription(
+            name="Right Seat Vent Status",
```
Comment (tmack8001) [inline custom_components/rivian/const.py:419]: Do we want to name these "Front Right Seat..." To be aligned with the naming of "Front Right Door/Window"?

**Review by tmack8001 — COMMENTED** [review]
Overall the additional info is nice thanks for adding those (bet more remote climate controls are coming soon #WinterIsComing)

One comment on the human name of the front seat sensors.

Regarding the previous json, if there is a server disconnect (HA not online, etc) wouldn't this incorrectly report the previous sensor states?

File: custom_components/rivian/__init__.py [hunk 2]
```diff
@@ -159,9 +159,14 @@ async def _update_api_data(self):
...
             vijson = await vehicle_info.json()
+            _LOGGER.debug(vijson)
 
             vehicle_info_items = self.build_vehicle_info_dict(vijson)
-            return vehicle_info_items
+            if vehicle_info_items:
+                self._previous_vehicle_info_items = vehicle_info_items
+                return vehicle_info_items
+            else:
+                return self._previous_vehicle_info_items
```
Comment (patfreeman) [inline custom_components/rivian/__init__.py:169]: If we don't have the unavailable info in the data of the sensors, we should log these events so we know when it happens.

**Comment by jrgutier** [comment]
1)No I dont think so, it would do a time-out exception. This only triggers if the data returned is blank. 

2) Yeah you're right. I was also thinking if I should be making some of them binary sensors too.

File: custom_components/rivian/const.py (hunk 1 above)
Comment (jrgutier) [reply custom_components/rivian/const.py:419]: I'll change

File: custom_components/rivian/__init__.py (hunk 2 above)
Comment (tmack8001) [reply custom_components/rivian/__init__.py:169]: Good point and maybe dump what did come back into logs for future debugging?

File: custom_components/rivian/const.py (hunk 1 above)
Comment (tmack8001) [reply custom_components/rivian/const.py:419]: Too bad Rivian themselves didn't stay consistent in the naming of the actual sensor here.

File: custom_components/rivian/__init__.py (hunk 2 above)
Comment (bretterer) [reply custom_components/rivian/__init__.py:169]: What happens if it goes unavailable for 10 minutes (arbitrary time). Will it always use the old value?   

Should we set a, if unavailable 10 times, show as unavailable?

Comment (tmack8001) [reply custom_components/rivian/__init__.py:169]: to sum up the discord chat, this `else` statement is only triggered when the entire call to Rivian fails to return **any** sensor data vs simply omitting a single sensor data point.

**Comment by krazos** [comment]
Looking pretty solid!  I especially appreciate having the SOC limit sensor for purposes of some of my calculated template sensors.  Beats having to create an automation to guess the SOC limit and update an `input_select` helper based on the state of charge, etc. 

This version is generally working smoothly for me.  A couple of quick thoughts: 

(1) This PR changes the `entity_id` of the window calibration sensors, so the old sensors show up as `unavailable`.  Is it possible to clean things up by having the integration automatically remove the old sensors so that users who upgrade from previous versions do not have to manually delete them?

(2) Is there a way to determine whether the vehicle is an R1T or R1S, and not add irrelevant sensors (e.g., third row seat heat sensors for R1T, tonneau and gear tunnel sensors for R1S, etc.)?

**Comment by tmack8001** [comment]
> (2) Is there a way to determine whether the vehicle is an R1T or R1S, and not add irrelevant sensors (e.g., third row seat heat sensors for R1T, tonneau and gear tunnel sensors for R1S, etc.)?

To do this properly would likely require a switch over to the newer graphql API. In that world we could auto discover the vehicle by the owner's credentials provided during setup vs asking for a specific VIN associated to the account. Anything before that there is no "vehicle type" given, though could make intelligent guesses based on a few (as you named) sensor data points if it is a T vs S type.

**Comment by krazos** [comment]
> > (2) Is there a way to determine whether the vehicle is an R1T or R1S, and not add irrelevant sensors (e.g., third row seat heat sensors for R1T, tonneau and gear tunnel sensors for R1S, etc.)?
> 
> To do this properly would likely require a switch over to the newer graphql API. In that world we could auto discover the vehicle by the owner's credentials provided during setup vs asking for a specific VIN associated to the account. Anything before that there is no "vehicle type" given, though could make intelligent guesses based on a few (as you named) sensor data points if it is a T vs S type. 

Ah, got it. This might not be worth the hassle, depending on the timeline for migrating over to the fancier API, but in the meantime, I suppose the config flow could include a dropdown for vehicle type alongside the VIN textbox. That would also allow us to include at least a basic description of the vehicle in the Device info view: 

image

**Comment by tmack8001** [comment]
> I suppose the config flow could include a dropdown for vehicle type alongside the VIN textbox.

With the GraphQL API we will be able to get back meta data about the vehicles you are assigned a key to (not just the one you are the `primary-owner` for.

```
{
   "__typename":"UserVehicle",
   "id":"<unique-internal-identifier>",
   "owner":null,
   "roles":[
      "primary-owner" // driver (not sure of that role value)
   ],
   "name":null, // eventually we might be able to assign names?
   "vin":"<vin>",
   "vas":{
      "__typename":"UserVehicleAccess",
      "vasVehicleId":"<access-key>",
      "vehiclePublicKey":"<key>"
   },
   "vehicle":{
      "__typename":"Vehicle",
      "model":"R1T", // or R1S / RCV?
      "mobileConfiguration":{
         "__typename":"VehicleMobileConfiguration",
         "trimOption":{ ... adventure vs launch .... },
         "exteriorColorOption":{ ... exterior color entity ... },
         "interiorColorOption":{ ... interior color entity ... }
      },
      "vehicleState":{
         "__typename":"VehicleState",
         "supportedFeatures":[
            { ... list of SupportedFeatures, not much useful here yet... likely R2 will have different features than R1? ... }
         ]
      }
   },
}
```

A future problem to solve for sure. Including the ability to create multiple devices as part of https://github.com/bretterer/home-assistant-rivian/issues/20

**Review by bretterer — APPROVED** [review]

**Pull request merged**